tensorflow/contrib/igfs/kernels/igfs.cc

```diff
@@ -0,0 +1,319 @@
+            << "]";
+
+  return Status::OK();
+}
+
+Status IGFS::DeleteDir(const string &file_name) {
+  std::shared_ptr<IGFSClient> client = CreateClient();
+  string path = TranslateName(file_name);
+
+  CtrlResponse<HandshakeResponse> handshake_response(true);
+  TF_RETURN_IF_ERROR(client->Handshake(&handshake_response));
+
+  CtrlResponse<ListFilesResponse> list_files_response(false);
+  TF_RETURN_IF_ERROR(client->ListFiles(&list_files_response, path));
+
+  if (!list_files_response.res.entries.empty()) {
+    return errors::FailedPrecondition("Can't delete a non-empty directory");
+  } else {
+    CtrlResponse<DeleteResponse> del_response(false);
+    TF_RETURN_IF_ERROR(client->Delete(&del_response, path, true));
+  }
+
+  LOG(INFO) << "Delete dir completed successful [file_name=" << file_name
+            << "]";
+
+  return Status::OK();
+}
+
+Status IGFS::GetFileSize(const string &file_name, uint64 *size) {
+  std::shared_ptr<IGFSClient> client = CreateClient();
+  string path = TranslateName(file_name);
+
+  CtrlResponse<HandshakeResponse> handshake_response(true);
+  TF_RETURN_IF_ERROR(client->Handshake(&handshake_response));
+
+  CtrlResponse<InfoResponse> info_response(false);
+  TF_RETURN_IF_ERROR(client->Info(&info_response, path));
+
+  *size = info_response.res.file_info.length;
+
+  LOG(INFO) << "Get file size completed successful [file_name=" << file_name
+            << "]";
+
+  return Status::OK();
+}
+
+Status IGFS::RenameFile(const string &src, const string &dst) {
+  std::shared_ptr<IGFSClient> client = CreateClient();
+  string src_path = TranslateName(src);
+  string dst_path = TranslateName(dst);
+
+  if (FileExists(dst).ok()) DeleteFile(dst);
+
+  CtrlResponse<HandshakeResponse> handshake_response(true);
+  TF_RETURN_IF_ERROR(client->Handshake(&handshake_response));
+
+  CtrlResponse<RenameResponse> rename_response(false);
+  TF_RETURN_IF_ERROR(client->Rename(&rename_response, src_path, dst_path));
+
+  if (!rename_response.res.successful)
+    return errors::NotFound("File ", src_path, " not found");
+
+  LOG(INFO) << "Rename file completed successful [src=" << src
+            << ", dst=" << dst << "]";
+
+  return Status::OK();
+}
+
+Status IGFS::Stat(const string &file_name, FileStatistics *stats) {
+  std::shared_ptr<IGFSClient> client = CreateClient();
+  string path = TranslateName(file_name);
+
+  CtrlResponse<HandshakeResponse> handshake_response(true);
+  TF_RETURN_IF_ERROR(client->Handshake(&handshake_response));
+
+  CtrlResponse<InfoResponse> info_response(false);
+  TF_RETURN_IF_ERROR(client->Info(&info_response, path));
+
+  IGFSFile info = info_response.res.file_info;
+
+  *stats = FileStatistics(info.length, info.modification_time * 1000000,
+                          (info.flags & 0x1) != 0);
+
+  LOG(INFO) << "Stat completed successful [file_name=" << file_name << "]";
+
+  return Status::OK();
+}
+
+std::shared_ptr<IGFSClient> IGFS::CreateClient() const {
```

Comment:
Is the "shared"-ness ever used here? It looks like this could return a `unique_ptr`.

Reply:
Replaced by `unique_ptr`, done.